### PR TITLE
Feature/con 9 interpolation method

### PR DESCRIPTION
**Pull request by Ciliaclaij**
## Description
This PR adds a new functionality for gain scheduling. The function made can be used to interpolate between the PID values matched to the different gait types.

## Changes

* Interpolate function
* Test functions added

### CODE REVIEW

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
+
+import rospy
+
+
+def interpolate(current_gains, needed_gains, gradient, delta_t):
+    if len(current_gains) != len(needed_gains):
+        rospy.logerr("needed_gains and current_gains not equal length")
+        return current_gains
```
Comment (Ishadijcks) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: Throw an exception here instead of returning incorrect values

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    next_gains = [0] * len(current_gains)
+    for i in range(len(current_gains)):
+        if current_gains[i] > needed_gains[i]:
+            next_gains[i] = current_gains[i] - gradient * delta_t
+            if next_gains[i] < needed_gains[i]:
+                next_gains[i] = needed_gains[i]
```
Comment (Ishadijcks) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]:
```
next_gains = max(needed_gains[i], current_gains[i] - gradient * delta_t)
```

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        else:
+            next_gains[i] = current_gains[i] + gradient * delta_t
+            if next_gains[i] > needed_gains[i]:
+                next_gains[i] = needed_gains[i]
```
Comment (Ishadijcks) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: Same

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
```
Comment (Olavhaasie) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: this can be removed

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py (hunk 1 above)
Comment (Olavhaasie) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: It is better to not log anything from library methods, then the user can determine whether to log or not. Furthermore, You should return `None` or raise an exception, otherwise it is hard to tell whether the returned array is the interpolated version or the first argument.

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        needed = [4, 3, 4]
+        while input != output:
+            result = interpolate(current, needed, 1, 0.1)
+        self.assertEqual(result, [4, 3, 4])
+
+    def test_interpolate_list_length(self):
+        current = [2, 5, 6, 8]
+        needed = [4, 3, 4]
+        with self.assertRaises(TypeError):
+            interpolate(current, needed, 1, 0.1)
```
Comment (Olavhaasie) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: This should not succeed :thinking:

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py (hunk 2 above)
Comment (Deakensbubbles) [reply march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: next_gains[i] = max(needed_gains[i], current_gains[i] - gradient * delta_t)

Subtle difference, but works better. also, using this method as suggested by Isha is neater, but requires an extra check for gains that are equal to the previous gains.

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py (hunk 3 above)
Comment (Deakensbubbles) [reply march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: another subtle difference: use min() instead of max() ergo: next_gains[i] = min(needed_gains[i], current_gains[i] + gradient * delta_t)

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py (hunk 1 above)
Comment (Roelemans) [inline march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: It is better to just throw an exception here. When this happens there is a significant problem with how this function is used. In the code this function is used, one can catch this exception (i.e. with try and except https://www.pythonforbeginners.com/error-handling/python-try-and-except/) and prevent the exo from shutting down. Its probably best to make your own exception (@Olavhaasie?) see https://www.programiz.com/python-programming/user-defined-exception

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        current = [5, 5, 5]
+        needed = [3, 3, 3]
+        result = interpolate(current, needed, 1, 0.1)
+        self.assertEqual(result, [4.9, 4.9, 4.9])
+
+    def test_interpolate_up_and_down(self):
+        current = [2, 5, 6]
+        needed = [4, 3, 4]
+        while input != output:
+            result = interpolate(current, needed, 1, 0.1)
```
Comment (Roelemans) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: If this does not work because someone broke your code, this loop will be infinite. Maybe nice to change it so that it iterates a maximum number of times :innocent:

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 5 above)
Comment (Roelemans) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Nice error testing :+1: unfortunate that this test will fail :P, but if you change the code so that it throws an error as mentioned above, you will only need to change the type of error checked for here :)

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        needed = [4, 3, 4]
+        with self.assertRaises(TypeError):
+            interpolate(current, needed, 1, 0.1)
+
+
+    def test_interpolate_different_paths(self):
+        current = [1, 1, 1]
+        needed = [2, 5, 4]
+        while current != needed:
+            result = interpolate(current, needed, 1, 0.1)
```
Comment (Roelemans) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Same as above. Try to avoid infinite loops :)

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        self.assertEqual(result, [2, 5, 4])
+
+
```
Comment (Roelemans) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Lastly, maybe nice to add one more test where you use a gradient and a delta_t that are not 1 and 0.1? If you accidentally hard-code these values in your function, shit is gonna go down :)

**Comment by Ishadijcks** [comment]
> Nice work! Your code is probably better tested than any other part of the code.

What if both lists are empty, or if I don't pass lists but strings ;)

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 6 above)
Comment (JorisWeeda) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: and also "output" is not defined right?

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 8 above)
Comment (JorisWeeda) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: I agree, or put a check in the function to verify the input values.

File: march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py (hunk 1 above)
Comment (Olavhaasie) [reply march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: Probably best to define your own

Comment (Ciliaclaij) [reply march_gain_scheduling/src/march_gain_scheduling/one_step_linear_interpolation.py]: I am trying to implement this into the code, but I can't figure out how.
This was my first guess:
`try:
        for i in range(len(current_gains)):
            if current_gains[i] > needed_gains[i]:
                 next_gains[i] = max(needed_gains[i], current_gains[i] - gradient * delta_t)
            else:
                 next_gains[i] = min(needed_gains[i], current_gains[i] - gradient * delta_t)
except:
           ....  
`
The problem when using this implementation is that you cover the situation where current_gains is larger than needed_gains, but not the other way around. Because that implementation will not make the loop fail.. Or do I need to make a implementation like this:

File: march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import rospy
+
+class UnequalLengthError(Exception):
```
Comment (JorisWeeda) [inline march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py]: When creating a customized error it is nice to add some additional information:
```suggestion
class UnequalLengthError(Exception):
    def __init__(self, msg=None):
        """Error class for the unequal length situation

        :param msg:
            The message that needs to be displayed.
        """
        if msg is None:
            msg = 'current_gains and needed_gains do not have the same length' 
         super(UnequalLengthError, self).__init__(msg)
```

File: march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+    print "current_gains and needed_gains do not have the same length"
```
Comment (JorisWeeda) [inline march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py]: And instead of printing maybe logging this error is better.

File: march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py (hunk 9 above)
Comment (Ishadijcks) [reply march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py]: `UnequalLengthError` implies it is generic across all arrays, yet its message is specific for current and needed gains

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 11]
```diff
@@ -21,23 +21,21 @@ def test_interpolate_down(self):
...
-
+        self.assertRaises(Exception, interpolate, current, needed, 1, 0.1)
 
     def test_interpolate_different_paths(self):
         current = [1, 1, 1]
         needed = [2, 5, 4]
-        while current != needed:
-            result = interpolate(current, needed, 1, 0.1)
-        self.assertEqual(result, [2, 5, 4])
+        for i in range(30):
```
Comment (Ishadijcks) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: 30?

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 8 above)
Comment (Olavhaasie) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Perhaps also add a test for empty lists?

File: march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py (hunk 10 above)
Comment (Ciliaclaij) [reply march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py]: Removed print

File: march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py (hunk 9 above)
Comment (Ciliaclaij) [reply march_gain_scheduling/src/march_gain_scheduling/unequal_length_error.py]: I am going to leave the exception empty. A message is passed with more explanation when it is called.

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 8 above)
Comment (Ciliaclaij) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Added both tests :+1:

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 11 above)
Comment (Ciliaclaij) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Added explanation

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        needed = [3, 3, 3]
+        result = interpolate(current, needed, 1, 0.1)
+        self.assertEqual(result, [4.9, 4.9, 4.9])
+
+    def test_interpolate_up_and_down(self):
+        current = [2, 5, 6]
+        needed = [4, 3, 4]
+        for i in range(25):
+            current = interpolate(current, needed, 3, 0.1)
+        self.assertEqual(current, [4, 3, 4])
```
[the remark this answers is not in the document]
Comment (Ciliaclaij) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Added Explanation

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            interpolate(current, needed, -1, 0.1)
+
+    def test_interpolate_different_paths(self):
+        current = [1, 1, 1]
+        needed = [2, 5, 4]
+        for i in range(26):  # Loop until needed is reached and a bit further
+            current = interpolate(current, needed, 1, 0.2)
+        self.assertEqual(current, [2, 5, 4])
+
+    def test_interpolate_up(self):
```
Comment (Olavhaasie) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: This test name is the same as the first test (probably because you copied it :speak_no_evil:)
```suggestion
    def test_interpolate_empty_lists(self):
```

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        current = []
+        needed = []
+        result = interpolate(current, needed, 1, 0.1)
+        self.assertEqual(result, [])
+
+
+
+if __name__ == '__main__':
+    rosunit.unitrun(PKG, 'test_multiply', OneStepLinearInterpolationTest)
```
Comment (Olavhaasie) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: `PKG` is undefined, you can solve this by adding 
```python
PKG = 'march_gain_scheduling`
```
At the top of the file, like they do in the [examples](https://wiki.ros.org/unittest#Code-level_Python_Unit_Tests).

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py (hunk 13 above)
Comment (Ciliaclaij) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: Fixed

Comment (Roelemans) [reply march_gain_scheduling/test/one_step_linear_interpolation_test.py]: PS this was a mistake in our tutorials, it has now been fixed

File: march_gain_scheduling/test/one_step_linear_interpolation_test.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    def test_negative_value_error(self):
+        current = [2, 5, 6]
+        needed = [4, 3, 4]
+        with self.assertRaises(NegativeValueError):
+            interpolate(current, needed, -1, 0.1)
+
+    def test_interpolate_different_paths(self):
+        current = [1, 1, 1]
+        needed = [2, 5, 4]
+        for i in range(26):  # Loop until needed is reached and a bit further
```
Comment (Ishadijcks) [inline march_gain_scheduling/test/one_step_linear_interpolation_test.py]: For this test to have any meaning you would need to confirm that it is not equal in earlier loops
